src/lib.rs

```diff
@@ -165,28 +165,27 @@ impl Hasher for FxHasher {
 
     #[inline]
     fn finish(&self) -> u64 {
-        // Since we used a multiplicative hash our top bits have the most
-        // entropy (with the top bit having the most, decreasing as you go).
-        // As most hash table implementations (including hashbrown) compute
-        // the bucket index from the bottom bits we want to move bits from the
-        // top to the bottom. Ideally we'd rotate left by exactly the hash table
-        // size, but as we don't know this we'll choose 20 bits, giving decent
-        // entropy up until 2^20 table sizes. On 32-bit hosts we'll dial it
-        // back down a bit to 15 bits.
-
+        // If input entropy is only found in the upper bits we want to spread
+        // it across the bits, which we do using a folded multiply.
+        //
+        // Using K as the multiplier for the folded multiply here isn't fully
+        // justified, there are probably better constants but we'd like to avoid
+        // loading another constant and it seems good enough.
+        //
+        // We don't use multiply_mix here as on 32-bit our input consists of two
+        // 32-bit numbers and multiply-mix does extra steps to mix what it
+        // thinks is a full 2x64-bit input.
         #[cfg(target_pointer_width = "64")]
-        const ROTATE: u32 = 20;
-        #[cfg(target_pointer_width = "32")]
-        const ROTATE: u32 = 15;
-
-        self.hash.rotate_left(ROTATE) as u64
+        {
+            let full = (self.hash as u128).wrapping_mul(K as u128);
```

Comment:
with the nightly feature it could use `widening_mul`

Reply:
I really don't see the point since it does the same thing and we still have to support stable anyway.

Reply:
just for dogfooding, but sure, not a priority.